runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -835,67 +837,153 @@ J9::ValuePropagation::constrainRecognizedMethod(TR::Node *node)
+         // Determine whether the arraylength is known and whether a BNDCHK is required
          //
-         flags8_t flagsForTransform(isLoadFlattenableArrayElement ? ValueTypesHelperCallTransform::IsArrayLoad
-                                                                  : ValueTypesHelperCallTransform::IsArrayStore);
-         flagsForTransform.set(ValueTypesHelperCallTransform::InsertDebugCounter);
+         bool isBoundCheckRequired = true;
+         int32_t arrayLength = -1;
 
-         if (isStoreFlattenableArrayElement && !owningMethodDoesNotContainStoreChecks(this, node))
+         if (owningMethodDoesNotContainBoundChecks(this, node))
             {
-            // If storing to an array whose component type is or might be a value type
-            // and the value that's being assigned is or might be null, both a run-time
-            // NULLCHK of the value is required (guarded by a check of whether the
-            // component type is a value type) and an ArrayStoreCHK are required;
-            // otherwise, only the ArrayStoreCHK is required.
-            //
-            if ((isCompTypeVT != TR_no) && (storeValueConstraint == NULL || !storeValueConstraint->isNonNullObject()))
+            isBoundCheckRequired = false;
+            }
+         else if (arrayConstraint != NULL)
+            {
+            int32_t lowerBoundLimit;
+            int32_t upperBoundLimit;
+            int32_t elementSize;
+            bool isKnownObj;
+
+            getArrayLengthLimits(arrayConstraint, lowerBoundLimit, upperBoundLimit, elementSize, isKnownObj);
+
+            if (lowerBoundLimit == upperBoundLimit)
                {
-               flagsForTransform.set(ValueTypesHelperCallTransform::RequiresStoreCheck);
-               flagsForTransform.set(ValueTypesHelperCallTransform::RequiresNullValueCheck);
+               arrayLength = lowerBoundLimit;
                }
-            else
+
+            TR::VPConstraint *indexConstraint = getConstraint(indexNode, arrayRefGlobal);
+            if (indexConstraint != NULL)
                {
-               flagsForTransform.set(ValueTypesHelperCallTransform::RequiresStoreCheck);
+               if (indexConstraint->getLowInt() >= 0 && indexConstraint->getHighInt() < lowerBoundLimit)
+                  {
+                  isBoundCheckRequired = false;
+                  }
                }
             }
 
-         if (!owningMethodDoesNotContainBoundChecks(this, node))
+         if (isBoundCheckRequired)
             {
-            flagsForTransform.set(ValueTypesHelperCallTransform::RequiresBoundCheck);
+            callToTransform->_flags.set(ValueTypesHelperCallTransform::RequiresBoundCheck);
             }
 
-         _valueTypesHelperCallsToBeFolded.add(
-               new (trStackMemory()) ValueTypesHelperCallTransform(_curTree, node, flagsForTransform, NULL));
+         if (arrayLength >= 0)
+            {
+            callToTransform->_arrayLength = arrayLength;
+            }
```

Comment:
maybe make setting up `arrayLength` is more explicit that it's only required if `isBoundCheckRequired` is true.
```suggestion 
 
         if (isBoundCheckRequired)
            {
            callToTransform->_flags.set(ValueTypesHelperCallTransform::RequiresBoundCheck);

            if (arrayLength >= 0)
               {
               callToTransform->_arrayLength = arrayLength;
               }
            }
```

Reply:
That sounds reasonable